examples/tcprtt_sockops/tcprtt_sockops.c

```diff
@@ -0,0 +1,120 @@
+// +build ignore
+
+#include "common.h"
+
+#include "bpf_endian.h"
+#include "bpf_sockops.h"
+#include "bpf_tracing.h"
+#include "tcprtt_sockops.h"
+
+char __license[] SEC("license") = "Dual MIT/GPL";
+
+static inline void init_sk_key(struct bpf_sock_ops *skops, struct sk_key *sk_key) {
+	sk_key->local_ip4   = bpf_ntohl(skops->local_ip4);
+	sk_key->remote_ip4  = bpf_ntohl(skops->remote_ip4);
+	sk_key->local_port  = skops->local_port;
+	sk_key->remote_port = bpf_ntohl(skops->remote_port);
+}
+
+static inline void bpf_sock_ops_establish_cb(struct bpf_sock_ops *skops, u8 sock_type) {
+	int err;
+	struct sk_info sk_info = {};
+	// Only process IPv4 sockets
+	if (skops == NULL || skops->family != AF_INET)
+		return;
+
+	// Initialize the 4-tuple key
+	init_sk_key(skops, &sk_info.sk_key);
+	sk_info.sk_type = sock_type;
+
+	// Store the socket info in map using the 4-tuple as key
+	// We keep track of TCP connections in 'established' state
+	err = bpf_map_update_elem(&map_estab_sk, &sk_info.sk_key, &sk_info, BPF_NOEXIST);
+	if (err != 0) {
+		bpf_printk("Error in storing sk_key in map");
```

Comment:
I'd prefer not encouraging printk, since it logs to the global trace pipe. Maybe just return here and add a comment why this could happen?